backupccl: prevent OR from linking in virtual file pointing to empty key space

msbutler opened the pull request:
See individual commits.

msbutler committed:
backupccl: fail online restore without table wide spans

To prevent reading from a virtual file that points to empty key space, this
patch fails an online restore if any of the target key spans are not table
spans; for example, if a table had a dropped index.

Informs #122176

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/122918)
<!-- Reviewable:end -->

msbutler committed:
backupccl: assert online restore span entry contains all of its backup files

One way online restore can link a virtual file that points to empty key space
in the backing file is if the restoreSpanEntry does not fully contain one of
its backup files. This patch fails the link phase of online restore if this
condition is detected.

Epic: none

Release note: none

msbutler commented:
unrelated test flake

msbutler commented on pkg/ccl/backupccl/restore_online.go:
```diff
@@ -194,6 +194,9 @@ func sendAddRemoteSSTWorker(
 				}
+				if !entry.Span.Contains(file.BackupFileEntrySpan) {
+					return errors.AssertionFailedf("file span %s is not contained in restore span %s", file.BackupFileEntrySpan, entry.Span)
+				}
```
one thought: we don't really clean up linked ssts if the link phase fails partway through. I feel fine with this for private preview. Better to fail loudly then silently corrupt the db.

We could gen spans before the actual link phase but i dont feel like optimizing UX private preview restore rn.

dt replied:
We believe entry.Span == restoringSubspan, right? Should we check that too just to be sure?

msbutler replied:
done and done.

stevendanna reviewed (commented):
Just some minor nits.

A bit of a bummer to have to bring out such a big hammer, but it's better than the alternative.

msbutler commented:
TFTR!

bors r=dt
